src/lib.rs

```diff
@@ -301,6 +312,8 @@ impl<'rx, 'tx> Eth<'rx, 'tx> {
         f: F,
     ) -> Result<R, TxError> {
         let result = self.tx_ring.send(length, f);
+        //Wait for the memory to sync with the cache
+        asm::dsb();
```

Comment:
I don't think this is needed for M4s, but we do need a compiler fence in `demand_pool`.

Reply:
Also, why do you think we need a `dsb` instead of just a `dmb` ? `dmb` is generated by `core::atomic::fence` which would avoid calling a function.

Reply:
I tested dmb instead and it works just fine. 
Reading into dmb vs dsb it looks like dmb is more appropriate here.
Since it is now just a compiler fence I don't think there is any harm in keeping it for both F7 and F4.
I initially used dsb because it was suggested on this thread: https://community.st.com/s/question/0D50X00009Xkgff/stm32f746-ethernet-dma-transmit-problem

Reply:
DMB isn't a compiler fence, it needs to still be inserted into the instructions. That's not to say that `core::atomic::fence` wouldn't emit it, though, so we might still be able to not call `dsb()` explicitly if there's a `fence()` call already. There's no harm in having it for the F4 but it won't do anything useful.

Did you test having neither dmb nor dsb and observe errors (i.e. are you testing with the DMA memory inside a cache-enabled region with caches on)?

I've used DSB when fixing this error in the past but it's definitely a stronger barrier and it does seem that DMB would be enough to ensure the write to DMASR will wait until after the write to the descriptor is complete.

Reply:
Checked some of my notes from last time. If I remember correctly, not having a DSB instruction at this point caused the DMA engine to never restart properly on an F7. If you do a test with neither DMB nor DSB, do you observe the failure? I'd want to be sure we were seeing the failure with neither instruction before concluding that the DMB was enough.

The cache issue is separate: if the descriptors are in a cached region, you need explicit cache cleaning/invalidation before you can enable the DMA, which is probably out of scope for this PR (but might be worth documenting somewhere). The best solution is to ensure the descriptors and their buffers are not placed in a cached region, since the caching doesn't help you anyway in those situations (and ideally they go into SRAM2 which can be separately mastered by the ethernet DMA).

Reply:
With no barrier at that location if I try to ping my processor I see it respond to every other ping request and timeout on the others.
Looking with wireshark, the STM was not sending anything out for the first ping, then would send out both the stale response and the new response on the second ping.
Putting either a dmb or a dsb at that location causes the STM to respond to every ping request properly.

I only saw this with release builds at higher HCLK speeds on my NUCLEO-F767ZI. 54Mhz, 108MHz, and 216Mhz all showed this behavior.
I don't have any F4s to test against, but I would be curious to know how the F4s behave with a release build and 168MHz HCLK.

Reply:
No barrier:
![image](https://user-images.githubusercontent.com/37219309/87042213-5b219d80-c1c1-11ea-9dbf-2e6c151a42ee.png)
DMB:
![image](https://user-images.githubusercontent.com/37219309/87042532-c9666000-c1c1-11ea-9a05-577b8741972b.png)

Reply:
Yep, that makes sense, thanks for testing! Looks like the DMB is the way to go then. It might be worth seeing if `core::sync::atomic::fence()` (perhaps with ordering release) will generate a DMB instruction for us, which would be lower overhead as the compiler could inline it without having to make a function call to cortex-m's external dmb function.

I'm confident the F4 doesn't require the DMB even at 168MHz HCLK, so I'm happy either way on whether this barrier is only included for F7 or is universally included.

Reply:
@adamgreig With the F7, a `DMB` will be needed because of the [store-buffer](https://developer.arm.com/documentation/ddi0489/b/memory-system/l1-caches/store-buffer), this was also observed in https://github.com/quartiq/stabilizer/issues/33. w.r.t invalidating the cache, I think a `DMB` should be enough, no ? At least it seems to be enough for the store-buffer.

For the F4s, I also don't think there's a need for a real fence, but we sure need some `compiler_fences` in some places. Also, `core::sync::atomic::fence` does generate a `DMB` instruction and we should probably use it instead, but I guess we should merge this with the `asm::dmb` for now, I'm planning on doing a follow up PR to inspect the code for places where we should have fences and add them, so I will change it for `atomic::fence` or `atomic::compiler_fence` depending on the core, what do you think ?

Reply:
I agree the DMB is both necessary and sufficient for the store buffer; when I mentioned the cache I mean the dcache (which is not cleaned by a DMB); if the dcache was enabled for the region the descriptor (or buffer) is in, it would need cleaning at this point (and the buffers would need invalidating on packet reception). Better to require that the descriptors and buffers are not placed in a dcache-enabled region, I think, and perhaps suggest they're placed in SRAM2 to reduce bus contention during DMA transfers.

Agreed on the compiler_fence. I think your plan sounds good.